bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,401 @@
+        let fingerprint_err = client_hello
+            .fingerprint_string_with_size(FingerprintType::JA3, expected_string.len() as u32 - 1)
+            .unwrap_err();
+        assert_eq!(fingerprint_err.kind(), ErrorType::UsageError);
+
+        let string = client_hello
+            .fingerprint_string_with_size(FingerprintType::JA3, expected_string.len() as u32)?;
+        assert_eq!(string, expected_string);
+        Ok(())
+    }
+
+    #[test]
+    fn invalid_client_bytes() {
+        let raw_client_hello_bytes =
+            "random_value_that_is_unlikely_to_be_valid_client_hello".as_bytes();
+        let result = ClientHello::parse_client_hello(raw_client_hello_bytes);
+        assert!(result.is_err());
+    }
+
+    // known value test case copied from s2n_fingerprint_ja3_test.c
+    #[test]
+    fn valid_client_bytes() {
+        let raw_client_hello = vec![
+            0x01, 0x00, 0x00, 0xEC, 0x03, 0x03, 0x90, 0xe8, 0xcc, 0xee, 0xe5, 0x70, 0xa2, 0xa1,
+            0x2f, 0x6b, 0x69, 0xd2, 0x66, 0x96, 0x0f, 0xcf, 0x20, 0xd5, 0x32, 0x6e, 0xc4, 0xb2,
+            0x8c, 0xc7, 0xbd, 0x0a, 0x06, 0xc2, 0xa5, 0x14, 0xfc, 0x34, 0x20, 0xaf, 0x72, 0xbf,
+            0x39, 0x99, 0xfb, 0x20, 0x70, 0xc3, 0x10, 0x83, 0x0c, 0xee, 0xfb, 0xfa, 0x72, 0xcc,
+            0x5d, 0xa8, 0x99, 0xb4, 0xc5, 0x53, 0xd6, 0x3d, 0xa0, 0x53, 0x7a, 0x5c, 0xbc, 0xf5,
+            0x0b, 0x00, 0x1e, 0xc0, 0x2b, 0xc0, 0x2f, 0xcc, 0xa9, 0xcc, 0xa8, 0xc0, 0x2c, 0xc0,
+            0x30, 0xc0, 0x0a, 0xc0, 0x09, 0xc0, 0x13, 0xc0, 0x14, 0x00, 0x33, 0x00, 0x39, 0x00,
+            0x2f, 0x00, 0x35, 0x00, 0x0a, 0x01, 0x00, 0x00, 0x85, 0x00, 0x00, 0x00, 0x23, 0x00,
+            0x21, 0x00, 0x00, 0x1e, 0x69, 0x6e, 0x63, 0x6f, 0x6d, 0x69, 0x6e, 0x67, 0x2e, 0x74,
+            0x65, 0x6c, 0x65, 0x6d, 0x65, 0x74, 0x72, 0x79, 0x2e, 0x6d, 0x6f, 0x7a, 0x69, 0x6c,
+            0x6c, 0x61, 0x2e, 0x6f, 0x72, 0x67, 0x00, 0x17, 0x00, 0x00, 0xff, 0x01, 0x00, 0x01,
+            0x00, 0x00, 0x0a, 0x00, 0x0a, 0x00, 0x08, 0x00, 0x1d, 0x00, 0x17, 0x00, 0x18, 0x00,
+            0x19, 0x00, 0x0b, 0x00, 0x02, 0x01, 0x00, 0x00, 0x23, 0x00, 0x00, 0x00, 0x10, 0x00,
+            0x0e, 0x00, 0x0c, 0x02, 0x68, 0x32, 0x08, 0x68, 0x74, 0x74, 0x70, 0x2f, 0x31, 0x2e,
+            0x31, 0x00, 0x05, 0x00, 0x05, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0d, 0x00, 0x18,
+            0x00, 0x16, 0x04, 0x03, 0x05, 0x03, 0x06, 0x03, 0x08, 0x04, 0x08, 0x05, 0x08, 0x06,
+            0x04, 0x01, 0x05, 0x01, 0x06, 0x01, 0x02, 0x03, 0x02, 0x01, 0x00, 0x1c, 0x00, 0x02,
+            0x40, 0x00,
+        ];
+        let expected_fingerprint = "771,49195-49199-52393-52392-49196-49200-\
+                                    49162-49161-49171-49172-51-57-47-53-10,0-\
+                                    23-65281-10-11-35-16-5-13-28,29-23-24-25,0";
+        let expected_hash_hex = "839bbe3ed07fed922ded5aaf714d6842";
+        known_test_case(raw_client_hello, expected_fingerprint, expected_hash_hex).unwrap();
+    }
```

Comment:
Rather than hard-coding a client hello here, could we just use our own s2n-tls client hello? Remember, we're tryign to test the bindings, not the underlying C implementation.

Same for the other tests-- I'd try to keep the tests focused on the bindings.

Reply:
> we just use our own s2n-tls client hello

I don't think I'm following this. Are you saying that we'd just handshake, dump the client hello to bytes, and then re-parse it? And then just not have any asserts for the contexts of the hash/fingerprint string?

> I'd try to keep the tests focused on the bindings.

I was definitely trying to keep with that goal
- test to make sure that IO stuff functions (the connection -> client hello stuff was a little hairy, want to make sure that's working)
- test to make sure that parsing function correctly (necessary to just make sure that we are getting correct results)
- test to make sure that broken parsing functions correctly (bindings have different error handling flows, want a test for that)
- test to look at debug implementation

Are there any that you think I should whittle down? I can shift to the "handshake then dump bytes" to get the client_hello for the debug test, but I don't think I'm comfortable removing the "known value" test.